info: Fix program version that recorded the data

qpakzk opened the pull request:
Solve #343 .

The `program version` showed uftrace version of currently running `uftrace info`.

Fix uftrace version that recorded the data.

When recording the data, write uftrace version in `uftrace.data/info` file. So It is possible to show a correct uftrace version by reading `uftrace.data/info` file.

honggyukim commented:
Hi @qpakzk, thanks for doing this. This change works fine but it breaks the previously recorded data. So it has to be changed in a backward compatible way.

qpakzk commented:
@honggyukim I see. But I have no idea what a backward compatible way is.

namhyung commented:
I don't think it breaks old data - it just skips the version info.  If there's a script which relies on the output format of uftrace info, it could break but I think it's extrememly rare..

Anyway we could print current version if there's no version info.  But I don't know it's better than printing nothing.  What do you guys think?

namhyung commented on uftrace.h:
```diff
@@ -72,6 +72,7 @@ enum uftrace_feat_bits {
 };
 
 enum uftrace_info_bits {
+	UFTRACE_VER,
```
Please make it simply `VERSION` unless there's a name conflict.  Or maybe we could rename them all to have `INFO_` prefix.

honggyukim commented:
I know that it works fine as follows:
```
$ uftrace record hello
Hello World!
$ uftrace info
# system information
# ==================
# program version     : v0.8.3-75-g5862
# recorded on         : Mon Apr 23 11:31:58 2018
# cmdline             : uftrace record a.out
# cpu info            : Intel(R) Xeon(R) CPU E5-2630 v4 @ 2.20GHz
# number of cpus      : 40 / 40 (online / possible)
# memory info         : 17.2 / 62.5 GB (free / total)
# system load         : 5.11 / 5.19 / 5.17 (1 / 5 / 15 min)
# kernel version      : Linux 4.4.0-116-generic
# hostname            : intel20
# distro              : "Ubuntu 16.04.3 LTS"
#
# process information
# ===================
# number of tasks     : 1
# task list           : 129393(a.out)
# exe image           : /home/honggyu/work/uftrace/git/final/hello/a.out
# build id            : 9f5c8583fa9704261fe9b620d96751787629fb81
# pattern             : regex
# exit status         : exited with code: 0
# elapsed time        : 0.005007996 sec
# cpu time            : 0.004 / 0.000 sec (sys / user)
# context switch      : 2 / 1 (voluntary / involuntary)
# max rss             : 3368 KB
# page fault          : 0 / 198 (major / minor)
# disk iops           : 0 / 16 (read / write)
```

But please see the below. I first recorded the data with current master.
```
$ cd uftrace
$ git checkout master
$ make install
$ uftrace record hello
Hello World!
```
Then if I read it with this new change, it shows the below error.
```
$ git checkout version
$ make install
$ uftrace info
uftrace: /home/honggyu/work/uftrace/utils/data-file.c:438:open_data_file
 ERROR: cannot read uftrace header info!
```
It's because there's no `uftrace_version` field in the old record data.

namhyung commented:
Ah, right.  The new bit should be added at last.

namhyung commented:
@qpakzk you should run tests before sending PR.

qpakzk commented:
@honggyukim Thank you for describing more details about the problem. 
@namhyung I'm sorry for careless tests.

I'm trying to print current version if there's no version info. And rename the enumeration constants' name of uftrace_info_bits with having `INFO_` prefix.

namhyung commented:
@qpakzk did you confirm that this change doesn't introduce new failures?

qpakzk commented:
@namhyung After I updated it, I tested `uftrace info` at v0.8.3 as well as v0.7. And it doesn't have new failures.
I added new data `uftrace version` at the bottom of uftrace.data/info, so it is compatible with old versions.

But it is necessary to modify `fill_uftrace_version()` function and `read_uftrace_version()` function after PR #348 is merged.

namhyung commented:
How did you test it with old versions?  And I'm afraid that it mightbreak some dump testcases.

namhyung commented:
@qpakzk any update?

qpakzk commented:
@namhyung I'm sorry to be late, but I had mid-term exams in school until today. I'll try to update!

qpakzk commented:
I'm sorry to update late. I reflected @honggyukim 's advice. 

```
$ ./runtest.py 097
Test case                 pg             finstrument-fu
------------------------: O0 O1 O2 O3 Os O0 O1 O2 O3 Os
097 dump_basic          : OK OK OK OK OK OK OK OK OK OK

$ ./runtest.py 098
Test case                 pg             finstrument-fu
------------------------: O0 O1 O2 O3 Os O0 O1 O2 O3 Os
098 dump_tid            : OK OK OK OK OK OK OK OK OK OK

$ ./runtest.py 099
Test case                 pg             finstrument-fu
------------------------: O0 O1 O2 O3 Os O0 O1 O2 O3 Os
099 dump_filter         : OK OK OK OK OK OK OK OK OK OK

$ ./runtest.py 100
Test case                 pg             finstrument-fu
------------------------: O0 O1 O2 O3 Os O0 O1 O2 O3 Os
100 dump_depth          : OK OK OK OK OK OK OK OK OK OK
```

And  following @honggyukim 's opinion, not to show version info when the data is recorded with old uftrace.

This is the old uftrace data.
```
$ uftrace info
# system information
# ==================
# recorded on         : Tue May 22 20:01:00 2018
# cmdline             : uftrace record test
# cpu info            : Intel(R) Core(TM) i7-4770HQ CPU @ 2.20GHz
# number of cpus      : 1 / 1 (online / possible)
# memory info         : 71.5 / 962.0 MB (free / total)
# system load         : 0.03 / 0.08 / 0.09 (1 / 5 / 15 min)
# kernel version      : Linux 4.16.0-041600-generic
# hostname            : frodo-hong-virtual-machine
# distro              : "Ubuntu 18.04 LTS"
```

This is the new uftrace data.
```
$ uftrace info
# system information
# ==================
# program version     : v0.8.3-133-g54e4
# recorded on         : Tue May 22 20:42:49 2018
# cmdline             : uftrace record test
# cpu info            : Intel(R) Core(TM) i7-4770HQ CPU @ 2.20GHz
# number of cpus      : 1 / 1 (online / possible)
# memory info         : 78.5 / 962.0 MB (free / total)
# system load         : 0.21 / 0.11 / 0.14 (1 / 5 / 15 min)
# kernel version      : Linux 4.16.0-041600-generic
# hostname            : frodo-hong-virtual-machine
# distro              : "Ubuntu 18.04 LTS"
```

honggyukim commented:
@qpakzk Thanks for doing this work!
